compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/Utils.desktop.kt

```diff
@@ -109,14 +110,40 @@ internal open class JLayeredPaneWithTransparencyHack: JLayeredPane() {
 }
 
 /**
- * Calls [block] synchronously on the event dispatching thread. If the calling thread is already the
- * event dispatch thread, simply executes [block]; otherwise schedules [block] to run on event
- * dispatching thread and waits for it to return.
+ * The set of the keys that have been scheduled to run on the EDT thread.
  */
-internal fun runOnEDTThread(block: () -> Unit) {
+private val scheduledKeys = Collections.synchronizedSet(mutableSetOf<Any>())
+
+/**
+ * Calls [block] on the event dispatching thread. If the calling thread is already the event
+ * dispatching thread, simply executes [block]; otherwise schedules [block] to run on event
+ * dispatching thread.
+ *
+ * @param key if not `null`, and a block with the same key has already been scheduled but not yet
+ * executed, the new block will not be scheduled.
+ */
+internal fun runOrScheduleOnEDTThread(key: Any? = null, block: () -> Unit) {
```

Comment:
1. a key API is usually more complicated and error prone. What if we create a class?

```
internal class EDTSingleExecutor {
    private var isScheduled = java.util.concurrent.atomic.AtomicBoolean(false)

    fun runOrSchedule(block: () -> Unit) {
        if (EventQueue.isDispatchThread()) {
            isScheduled.set(false)
            block()
        } else {
            if (isScheduled.getAndSet(true)) {
                return
            }
            EventQueue.invokeLater {
                if (isScheduled.getAndSet(false)) {
                   block()
                }
            }
        }
    }
}
```

2. "null" key check slightly bloats the code, and doesn't seem needed. I would remove nullability if we won't do the first.

Reply:
> a key API is usually more complicated and error prone

Besides this, there can be 2 concurrent scenes. We need to have different keys for them

Reply:
> Besides this, there can be 2 concurrent scenes. We need to have different keys for them

Yes, I hadn't considered that.